pyccel/parser/syntax/openmp.py

```diff
@@ -11,396 +11,962 @@
-                txt = '{0} {1}'.format(txt, clause.expr)
-            else:
-                raise TypeError('Wrong clause for OmpSingleConstruct')
+class OmpCancelConstruct(OmpConstruct):
+    """Class representing a Cancel construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Cancel_Construct, (OmpCancelType,), **kwargs)
+
+class OmpTargetConstruct(OmpConstruct):
+    """Class representing a Target construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Target_Construct, _valid_target_clauses, **kwargs)
+
+class OmpTeamsConstruct(OmpConstruct):
+    """Class representing a Teams construct."""
+    def __init__(self, **kwargs):
+        super().__init__(OMP_Teams_Construct, _valid_teams_clauses, **kwargs)
 
-        return OMP_Single_Construct(txt)
+class OmpAtomicConstruct(OmpConstruct):
+    """Class representing an Atomic construct ."""
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, _valid_atomic_clauses, **kwargs)
 
 class OmpEndClause(BasicStmt):
-    """Class representing a ."""
+    """Class representing an end construct."""
     def __init__(self, **kwargs):
-        """
-        """
         self.construct = kwargs.pop('construct')
         self.simd      = kwargs.pop('simd', '')
         self.nowait    = kwargs.pop('nowait', '')
 
-        super(OmpEndClause, self).__init__(**kwargs)
+        construct = ' '.join(self.construct)
+        txt = 'end {0} {1} {2}'.format(construct, self.simd, self.nowait)
+
+        self._expr = Omp_End_Clause(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        return self._expr
+
+class OmpFinal(BasicStmt):
+    """Class representing a final clause"""
+    def __init__(self, **kwargs):
+        self.final = kwargs.pop('final')
+
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
-            print("> OmpEndClause: expr")
+            print("> {}: expr".format(type(self).__name__))
 
-        txt = 'end {0} {1} {2}'.format(self.construct, self.simd, self.nowait)
-        return Omp_End_Clause(txt)
+        final = self.final
+        return 'final({})'.format(final)
 
 class OmpNumThread(BasicStmt):
-    """Class representing a ."""
+    """Class representing a num_thread clause."""
     def __init__(self, **kwargs):
-        """
-        """
         self.thread = kwargs.pop('thread')
 
-        super(OmpNumThread, self).__init__(**kwargs)
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         # TODO check if variable exist in namespace
         if DEBUG:
-            print("> OmpNumThread: expr")
+            print("> {}: expr".format(type(self).__name__))
 
         thread = self.thread
         return 'num_threads({})'.format(thread)
 
+class OmpNumTeams(BasicStmt):
+    """Class representing a num_teams clause."""
```

Comment:
num_teams -> NumTeams.

Reply:
there is no `NumTeams` in openmp, this class is representing a `num_teams` clause

Reply:
i think that `num_teams` for code writing and `NumTeams` for the docstring.

Reply:
`num_theam` is the name of the clause in openmp, there is nothing named `NumTeams` in openmp, and we don't use this name in pyccel, so why would we use it ?